pkg/sendbuffer/sendbuffer.go

```diff
@@ -0,0 +1,264 @@
+		return 0, nil
+	}
+
+	// if the single data piece is larger than the max send size, drop it and log
+	if len(data) > sb.maxSendSize {
+		sb.logger.Log(
+			"msg", "dropped data because element greater than max send size",
+			"size_of_data", len(data),
+			"max_send_size", sb.maxSendSize,
+			"head", string(data)[0:minInt(len(data), 100)],
+		)
+		return len(data), nil
+	}
+
+	// if we are full, something has backed up
+	// purge everything
+	if len(data)+sb.size > sb.maxSize {
+		sb.logger.Log(
+			"msg", "reached capacity, dropping all data and starting over",
+			"size_of_data", len(data),
+			"buffer_size", sb.size,
+			"size_plus_data", sb.size+len(data),
+			"max_size", sb.maxSize,
+		)
+
+		if err := sb.purge(len(sb.dataKeys)); err != nil {
+			// something has gone horribly wrong
+			sb.logger.Log("msg", "failed to purge", "err", err)
+			return len(data), nil
+		}
+
+		sb.size = 0
+	}
+
+	dataKey := []byte(ulid.New())
+	if err := sb.kvStore.Set(dataKey, data); err != nil {
+		// log it and move on
+		sb.logger.Log("msg", "failed to store data", "err", err)
+		return len(data), nil
+	}
+
+	sb.dataKeys = append(sb.dataKeys, dataKey)
+	sb.size += len(data) + len(dataKey)
+	return len(data), nil
+}
+
+func (sb *SendBuffer) StartSending() {
+	if sb.isSending {
+		return
+	}
+
+	sb.isSending = true
+
+	go func() {
+		ticker := time.NewTicker(sb.sendInterval)
+		defer ticker.Stop()
+
+		for {
+			if err := sb.sendAndPurge(); err != nil {
+				sb.logger.Log("msg", "failed to send and purge", "err", err)
+			}
+
+			select {
+			case <-ticker.C:
+				continue
+			case <-sb.stopSendingChan:
+				sb.isSending = false
+				return
+			}
+		}
+	}()
+}
+
+func (sb *SendBuffer) StopSending() {
+	select {
+	case sb.stopSendingChan <- struct{}{}:
+	default:
+	}
+}
+
+func (sb *SendBuffer) sendAndPurge() error {
+	sb.writeMutex.Lock()
+	defer sb.writeMutex.Unlock()
+
+	if len(sb.dataKeys) == 0 {
+		return nil
+	}
+
+	removeDataKeysToIndex := 0
+	sendBuffer := &bytes.Buffer{}
+	bytesToDelete := 0
+
+	for _, dataKey := range sb.dataKeys {
+		data, err := sb.kvStore.Get(dataKey)
+		if err != nil {
+			sb.logger.Log("msg", "could not fetch data from kv store", "data_key", dataKey)
+			removeDataKeysToIndex++
+			continue
+		}
+
+		// ready to send, break from loop
+		if len(data)+sendBuffer.Len() > sb.maxSendSize {
+			break
+		}
+
+		// add to buffer to be sent
+		if _, err := sendBuffer.Write(data); err != nil {
+			return fmt.Errorf("writing data to buffer: %w", err)
+		}
+
+		bytesToDelete += len(data) + len(dataKey)
+		removeDataKeysToIndex++
+	}
+
+	if !sb.sendMutex.TryLock() {
+		sb.logger.Log("msg", "could not get lock on send mutex, will retry")
+		return nil
+	}
+
+	// have lock on write mutex and send mutex at this point
+	if err := sb.purge(removeDataKeysToIndex); err != nil {
+		sb.sendMutex.Unlock()
+		return fmt.Errorf("purging data from kv store: %w", err)
+	}
+
+	sb.size -= bytesToDelete
+
+	// send off async
+	go func() {
```

Comment:
`sendAndPurge` is already in a goroutine. Why spawn another?

Reply:
I suppose it's easy to add it in later if we decided that we want to back with on disk KVStore, I can drop it for simplicity sake

Reply:
I did this so that the write mutex does not have to wait on the send

Reply:
busted out the flushing to another func with it's own write lock, so no more go routine, looks cleaner